src/transformer/transformer.js

```diff
@@ -58,43 +59,43 @@ export default async function (transformCallback) {
 	async function startTransformation() {
 		let connection;
 		let channel;
+		let hasFailed = false;
+		const pendingPromises = [];
+		let counter = -1;
+		let numberOfRecords = 0;
 
 		const ApiClient = createApiClient({url: API_URL, username: API_USERNAME, password: API_PASSWORD, userAgent: API_CLIENT_USER_AGENT});
 		const {readStream} = await ApiClient.getBlobContent({id: BLOB_ID});
+		let TransformClient;
 
 		logger.log('info', `Starting transformation for blob ${BLOB_ID}`);
 
 		try {
 			connection = await amqplib.connect(AMQP_URL);
 			channel = await connection.createChannel();
-
-			const records = await transformCallback(readStream);
-			const failedRecords = records.filter(r => r.failed).map(result => {
-				return {...result, record: result.record.toObject(), timestamp: moment()};
-			});
-
-			logger.log('debug', `${failedRecords.length} records failed`);
-
-			await ApiClient.setTransformationDone({
-				id: BLOB_ID,
-				numberOfRecords: records.length,
-				failedRecords
-			});
+			TransformClient = transformCallback(readStream);
+
+			try {
+				await new Promise((resolve, reject) => {
+					TransformClient
```

Comment:
Transformer implementation should only emit 3 events:
- **record**: When a record is processed
- **end**: When all records have been processed
- **error**: When an error is thrown

Transformer will use it's own logger to log messages. Counter can be implemented in this module /commons/transformer) based on the **record**-event.

Reply:
Internal counter is needed to prevent too early channel closing